### ISSUE TITLE
suggestion: background shading of blocks in editor

### ISSUE THREAD

**Issue by ericscheid** [naturalcrit/homebrewery#1845]
From https://www.reddit.com/r/homebrewery/comments/qum9t8/suggestion_would_love_if_the_editor_automatically/

> I sometimes find myself getting lost/confused in piles of text, especially when scrolling through a large brew. Having some sort of subtle highlight which makes tables and notes different from other text would be awesome for readability. Ideally, having different types of tables highlight with their own color would be awesome (stat blocks in yellow, notes in purple or green, tables in green or orange, etc.) but even just one color for all of them would be helpful.

![image](https://user-images.githubusercontent.com/2605467/142620446-6851173d-43f1-4bc0-86f7-24cef20b58c5.png)

**Comment by ericscheid** [comment]
So, blocks that could be tinted in the editor are:

1. `{{` curly blocks and spans `}}`
2. `{{` curly blocks and spans with specific class names picking up the background from the css `}}`
3. `| tables |`
4. `> blockquotes`
5. `<direct-html-elements>` (but deprecated)
6. <code>&#96;&#96;&#96;code blocks&#96;&#96;&#96;</code> and <code>&#96;code spans&#96;</code>

Of these, # 2 might be tricky. Note that # 4 already has styling in the editor of making the text green (but the resulting `<blockquote>` html code that is generated doesn't have any css styling in the V3 phb53style.css file (it does in legacy render style).

**Comment by Gazook89** [comment]
> Of these, # 2 might be tricky. Smexy as heck, but very tricky to implement. That said, we could trap for specific known classnames (e.g. {{monster or {{note) and hard-code specific backgrounds.

I think trapping for specific class names is *good enough* at least to start-- otherwise you'd have to find all the class names, then search the entire document + the style text for those style declarations, and then figure out which has precedence.

So if looking for specific class names, should it look only at the first position?  `{{note ...`  would apply the background for notes, but `{{cover,note, ...` would not?    Another option is to search the whole line for specific class names, but then you run into issues if multiple "special" class names appear....just set the background based on the first match?

At this point I'm leaning towards using the first position only.  So div blocks that start with the below patterns would trigger special styling:

- `{{note `  -- light green
- `{{descriptive `  -- light brown
- `{{spellList `  -- light blue
- `{{monster `  -- light orange
- `{{toc `  -- light yellow

There are other block snippets that aren't included (such as .wide) but I'm not sure they need to have special backgrounds?  

Also, there is a choice of adding the highlight into the line number gutter of the editor, rather than behind the text (or, in addition to behind the text).   I don't think we would want to do this, but it's an option.

**Comment by calculuschild** [comment]
I imagine we could do this with some clever CSS variables in the theme files so that each class can change the highlighting depending on the theme.

My instinct is we should wait on getting Themes (and CodeMirror themes) working first and make sure we have a clear consensus before putting effort into a PR.

**Comment by dbolack-ab** [comment]
Is this still an idea going forward?

**Comment by 5e-Cleric** [comment]
> Is this still an idea going forward?

see last comment on #1855

**Comment by dbolack-ab** [comment]
Feels like a no.

**Comment by ericscheid** [comment]
> > Is this still an idea going forward?
> 
> see last comment on #1855 

vis..

> Let's have further discussions on #1845 so we can lay out a specific set of behaviors we want this feature to have. This PR looks promising but I think we need to develop the idea a bit more first.

I'm still interested, I just don't have any further suggestions of additional behaviours to roll in to the mvp.

Could we do a PR of the MVP of this feature request, and refactor new behaviours in following issues/PRs? I'm seeing lots of ideas/suggestions die on the vine because the basic idea gets bogged down in complications from extensions.

More MVPs, with iterations; fewer cathedrals.

**Comment by 5e-Cleric** [comment]
Considering how many people objected to the active line highlight, this wouldn't roll out well. Considering closing it.

**Comment by 5e-Cleric** [comment]
I call a vote.

Result: If positive 😄, opening up the PR for it. If negative 😕, closing the issue until further notice.

Remember to vote by reacting to this comment with the following emojis: 😄 for positive, 😕 for negative.

@naturalcrit/issue-voters

**Comment by ericscheid** [comment]
> This PR looks promising but I think 
> we need to develop the idea a bit more first.

I read this as *pause the PR, discuss in the issue*. Not "close the issue".

A legitimate concern has been raised as to how the community might react to the visual change.

We have a few screen caps (mockups or otherwise) .. post those to reddit and gauge the reaction.

**Comment by ericscheid** [comment]
BTW, it took me ~two~ _three_ goes to figure out which emoji to vote with.

Q: shall we close this issue?  
positive = no, don't close the issue  
negative = yes, close the issue

**Comment by 5e-Cleric** [comment]
> BTW, it took me two goes to figure out which emoji to vote with.
> 
> Q: shall we close this issue? positive = no, don't close the issue negative = yes, close the issue

The question being whether this issue should go forward or not, not whether we should close it.

**Comment by 5e-Cleric** [comment]
Working on this, worried that it might be a bit expensive since we got to check for blocks inside blocks inside blocks...

Also, what do we do with the colors when a block contains a block? do the colors mix? does the inner color prevail?

**Comment by ericscheid** [comment]
> Also, what do we do with the colors when a block contains a block? do the colors mix? does the inner color prevail?

My approach to that has usually been to define the colours with bold saturation but fade them out with opacity.

Like this: 
> `div {background-color: #ffc0cb06; /* pink, barely */ }` 

... that way the detection of nesting and the handling of colouring is fully offloaded to CSS.

(I just realised that this approach probably won't work with the editor, which has all the content sliced into lines with no nesting. Bah humbug.)

> <img width="1031" height="776" alt="divs have a mostly transparent pink tint" src="https://github.com/user-attachments/assets/3220f1d5-4d04-4117-95ce-32993dff26c0" />

**Comment by Gazook89** [comment]
Yeah, the editor has no idea what is nested and not nested without us manually looping through.

I think overall this is going to look terrible.  You'll need to find satisfactory colors, which sounds awful just on the surface.  Then, consider all of the themes.  You'll be looking to contrast harmoniously with background colors and every possible token highlight color. Then those colors have to blend or nest?

If pursuing that, I recommend getting comfy with the CSS _Relative Color_ syntax `background: hcl(from var(--blah) h c .9)` and setting every theme to have it's colors defined by custom variables.  Then, rather than hardcoding in arbitrary colors, you take the theme's colors and just tweak them (either subtly, or rotate around the color wheel by thirds for high contrast).  Will still require some experimenting but may be more flexible.

But otherwise....

What about options like the below for offering better editor navigation?

- [scrollbar minimap](https://www.npmjs.com/package/@replit/codemirror-minimap)
- [annotated scrollbar (link to Windows article about them generally)](https://learn.microsoft.com/en-us/windows/apps/develop/ui/controls/annotated-scrollbar) and [link to codemirror 5 add-on](https://codemirror.net/5/doc/manual.html#addon_annotatescrollbar)

**Comment by 5e-Cleric** [comment]
> Yeah, the editor has no idea what is nested and not nested without us manually looping through.
> 
> I think overall this is going to look terrible. You'll need to find satisfactory colors, which sounds awful just on the surface. Then, consider all of the themes. You'll be looking to contrast harmoniously with background colors and every possible token highlight color. Then those colors have to blend or nest?

Oh damn i had not considered themes....

> If pursuing that, I recommend getting comfy with the CSS _Relative Color_ syntax `background: hcl(from var(--blah) h c .9)` and setting every theme to have it's colors defined by custom variables. Then, rather than hardcoding in arbitrary colors, you take the theme's colors and just tweak them (either subtly, or rotate around the color wheel by thirds for high contrast). Will still require some experimenting but may be more flexible.

The relative color syntax would work... now that we have the themes in their own repo, yeah. Would require me to set that up but its not hard, just long and repetitive. And it might still look bad.


> But otherwise....
> 
> What about options like the below for offering better editor navigation?
> 
> * [scrollbar minimap](https://www.npmjs.com/package/@replit/codemirror-minimap)
> * [annotated scrollbar (link to Windows article about them generally)](https://learn.microsoft.com/en-us/windows/apps/develop/ui/controls/annotated-scrollbar) and [link to codemirror 5 add-on](https://codemirror.net/5/doc/manual.html#addon_annotatescrollbar)

Not opposed. 


Perhaps we should re-vote.
